libp2p-floodsub/src/lib.rs

```diff
@@ -0,0 +1,640 @@
+                                    future::Loop::Continue((floodsub_sink, rest))
+                                });
+                                Box::new(future) as Box<_>
+                            }
+
+                            None => {
+                                // Both the connection stream and `rx` are empty, so we break
+                                // the loop.
+                                trace!(target: "libp2p-floodsub", "Pubsub future clean finish");
+                                // TODO: what if multiple connections?
+                                inner.remote_connections.write().remove(&remote_addr);
+                                let future = future::ok(future::Loop::Break(()));
+                                Box::new(future) as Box<Future<Item = _, Error = _>>
+                            }
+                        }
+                    })
+            },
+        );
+
+        future::ok(FloodSubFuture {
+            inner: Box::new(future) as Box<_>,
+        })
+    }
+}
+
+/// Allows one to control the behaviour of the floodsub system.
+#[derive(Clone)]
+pub struct FloodSubController<T, C>
+where
+    T: MuxedTransport + 'static,                // TODO: 'static :-/
+    C: ConnectionUpgrade<T::RawConn> + 'static, // TODO: 'static :-/
+{
+    inner: Arc<Inner>,
+    swarm: SwarmController<T, C>,
+}
+
+struct Inner {
+    // Our local peer ID multihash, to pass as the source.
+    peer_id: Vec<u8>,
+
+    // Channel where to send the messages that should be dispatched to the user.
+    output_tx: mpsc::UnboundedSender<Message>,
+
+    // Active connections with a remote.
+    remote_connections: RwLock<FnvHashMap<Multiaddr, RemoteInfo>>,
+
+    // List of topics we're subscribed to. Necessary in order to filter out messages that we
+    // erroneously receive.
+    subscribed_topics: RwLock<Vec<Topic>>,
+
+    // Sequence number for the messages we send.
+    seq_no: AtomicUsize,
+
+    // We keep track of the messages we received (in the format `(remote ID, seq_no)`) so that we
+    // don't dispatch the same message twice if we receive it twice on the network.
+    // TODO: the `HashSet` will keep growing indefinitely :-/
+    received: Mutex<FnvHashSet<(Vec<u8>, Vec<u8>)>>,
+}
+
+struct RemoteInfo {
+    // Sender to send data over the socket to that host.
+    sender: mpsc::UnboundedSender<BytesMut>,
+    // Topics the remote is registered to.
+    subscribed_topics: RwLock<FnvHashSet<TopicHash>>,
+}
+
+impl fmt::Debug for Inner {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.debug_struct("Inner")
+            .field("peer_id", &self.peer_id)
+            .field(
+                "num_remote_connections",
+                &self.remote_connections.read().len(),
+            )
+            .field(
+                "num_subscribed_topics",
+                &self.subscribed_topics.read().len(),
```

Comment:
If we're locking `subscribed_topics` anyway, why not just print all the subscribed topics? Same with `remote_connections`. There's a lot of data but I'd rather show too much than too little. `RwLock` implements `Debug`, just use `derive(Debug)`.